File: latest/week1/workshop3_m5/W03_M05_demo.R
# ============================================================
# Workshop 3 — Module 5: Proportions — Pie, Donut, Treemap & Waffle
# R Demonstration Script — UNYT Tirana
# ============================================================
library(tidyverse); dir.create("output",showWarnings=FALSE)
apps <- read_csv("googleplaystore.csv") |> filter(Type %in% c("Free","Paid"),!is.na(Rating))
top6 <- apps |> count(Category,sort=TRUE) |> slice_max(n,n=6)

# 1. PIE (to show why it fails)
p_pie <- top6 |>
  ggplot(aes(x="",y=n,fill=Category)) +
  geom_col(width=1,color="white",linewidth=1) +
  coord_polar(theta="y") +
  scale_fill_brewer(palette="Set2") +
  theme_void() + labs(title="Pie: Hard to Compare 6 Slices")
ggsave("output/pie_fail.png",p_pie,width=5,height=5,dpi=300)

# 2. DONUT (KPI)
df_kpi <- tibble(cat=c("Free","Paid"),n=c(sum(apps$Type=="Free"),sum(apps$Type=="Paid")))
p_donut <- ggplot(df_kpi,aes(x=2,y=n,fill=cat)) +
  geom_col(width=1,color="white",linewidth=2) +
  coord_polar(theta="y") + xlim(0.5,2.5) +
  scale_fill_manual(values=c(Free="#1565C0",Paid="#EEEEEE")) +
  annotate("text",x=0.5,y=0,label=paste0(round(df_kpi$n[1]/sum(df_kpi$n)*100),"%"),
    size=12,fontface="bold",color="#1565C0") +
  annotate("text",x=0.5,y=sum(df_kpi$n)*0.05,label="Free Apps",size=4,color="#555") +
  theme_void() + theme(legend.position="none") +
  labs(title="Donut: Single KPI")
ggsave("output/donut_kpi.png",p_donut,width=5,height=5,dpi=300)

# 3. 100% STACKED BAR (the better alternative)
p_stacked <- top6 |>
  mutate(pct=n/sum(n)) |>
  ggplot(aes(x="Apps",y=pct,fill=fct_reorder(Category,n))) +
  geom_col(width=0.6,color="white",linewidth=1.5) +
  geom_text(aes(label=paste0(round(pct*100),"%")),
    position=position_stack(vjust=0.5),size=3,fontface="bold",color="white") +
  scale_fill_brewer(palette="Set2") +
  scale_y_continuous(labels=scales::percent) +
  coord_flip() + theme_minimal() +
  labs(title="100% Stacked Bar: Better Than Pie",fill="Category",y=NULL,x=NULL)
ggsave("output/stacked_bar.png",p_stacked,width=8,height=3,dpi=300)

# 4. TREEMAP
# library(treemapify)
# p_tree <- ggplot(top6, aes(area=n, fill=Category,
#     label=paste(Category,scales::comma(n),sep="\n"))) +
#   geom_treemap(color="white",size=2) +
#   geom_treemap_text(fontface="bold",color="white",place="centre") +
#   scale_fill_brewer(palette="Set2") + theme(legend.position="none") +
#   labs(title="Treemap: Area = Count")
# ggsave("output/treemap.png",p_tree,width=7,height=5,dpi=300)

# 5. WAFFLE
# library(waffle)
# waffle(c(FAMILY=38,GAME=20,TOOLS=18,PHOTO=8,PROD=8,BUS=8),
#   rows=10,colors=c("#1565C0","#E53935","#2E7D32","#E65100","#7B1FA2","#00695C"),
#   title="App Categories (%)")

cat("\n── All W03-M05 R plots saved ──\n")

File: latest/week1/workshop3_m5/W03_M05_demo.py
"""W03-M05: Proportions — Pie, Donut, Treemap & Waffle — Python — UNYT"""
import numpy as np, matplotlib.pyplot as plt, os
os.makedirs("output",exist_ok=True); np.random.seed(42)
cats=["FAMILY","GAME","TOOLS","PHOTOGRAPHY","PRODUCTIVITY","BUSINESS"]
vals=[1832,959,843,376,374,420]; colors=["#1565C0","#E53935","#2E7D32","#E65100","#7B1FA2","#00695C"]

# 1. PIE vs BAR
fig,(ax1,ax2)=plt.subplots(1,2,figsize=(11,4.5))
ax1.pie(vals,labels=cats,autopct="%1.0f%%",colors=colors,startangle=90,textprops={"fontsize":7})
ax1.set_title("Pie: 6 slices = hard to compare",fontweight="bold",color="#C62828")
idx=np.argsort(vals); sc=[cats[i] for i in idx]; sv=[vals[i] for i in idx]; cc=[colors[i] for i in idx]
ax2.barh(sc,sv,color=cc,height=0.55)
for i,v in enumerate(sv): ax2.text(v+15,i,f"{v:,} ({v/sum(vals)*100:.0f}%)",va="center",fontsize=7,fontweight="bold")
ax2.set_title("Bar + %: always better",fontweight="bold",color="#1565C0")
for sp in ["top","right","left"]: ax2.spines[sp].set_visible(False)
ax2.set_xticks([]); ax2.tick_params(left=False,labelsize=7)
fig.suptitle("Pie vs Bar",fontweight="bold"); plt.tight_layout()
plt.savefig("output/pie_vs_bar_py.png",dpi=300); plt.close()

# 2. DONUT KPI
fig,ax=plt.subplots(figsize=(5,4.5))
ax.pie([73,27],colors=["#1565C0","#EEEEEE"],startangle=90,wedgeprops={"width":0.35,"edgecolor":"white","linewidth":2})
ax.text(0,0.05,"73%",ha="center",va="center",fontsize=28,fontweight="bold",color="#1565C0")
ax.text(0,-0.12,"Free Apps",ha="center",va="center",fontsize=10,color="#555")
ax.set_title("Donut: Single KPI",fontweight="bold"); plt.tight_layout()
plt.savefig("output/donut_py.png",dpi=300); plt.close()

# 3. TREEMAP (squarify)
try:
    import squarify
    fig,ax=plt.subplots(figsize=(8,5))
    squarify.plot(sizes=vals,label=[f"{c}\n{v:,}" for c,v in zip(cats,vals)],color=colors,alpha=0.7,
        edgecolor="white",linewidth=2,text_kwargs={"fontsize":8,"fontweight":"bold","color":"white"},ax=ax)
    ax.axis("off"); ax.set_title("Treemap: Area = Count",fontweight="bold")
    plt.tight_layout(); plt.savefig("output/treemap_py.png",dpi=300); plt.close()
    print("  treemap saved (squarify)")
except ImportError:
    print("  squarify not installed — pip install squarify")

# 4. WAFFLE (manual)
fig,ax=plt.subplots(figsize=(6,5))
total=sum(vals); pcts=[round(v/total*100) for v in vals]; diff=100-sum(pcts); pcts[0]+=diff
grid=np.zeros((10,10),dtype=int); idx2=0; cat_idx=0
for i in range(10):
    for j in range(10):
        grid[9-i,j]=cat_idx; idx2+=1
        if idx2>=sum(pcts[:cat_idx+1]) and cat_idx<len(pcts)-1: cat_idx+=1
for i in range(10):
    for j in range(10):
        ax.add_patch(plt.Rectangle((j,i),0.9,0.9,facecolor=colors[grid[i,j]],edgecolor="white",linewidth=1.5))
ax.set_xlim(-0.2,10.2); ax.set_ylim(-0.2,10.2); ax.set_aspect("equal"); ax.axis("off")
for k,(cat,pct,c) in enumerate(zip(cats,pcts,colors)):
    ax.text(10.8,8.5-k*1.2,f"■ {cat} ({pct}%)",fontsize=8,color=c,fontweight="bold",va="center")
ax.set_title("Waffle: 1 square = 1%",fontweight="bold"); plt.tight_layout()
plt.savefig("output/waffle_py.png",dpi=300); plt.close()

# 5. 100% STACKED BAR
fig,ax=plt.subplots(figsize=(9,2.5)); left=0
for cat,val,color in zip(cats,vals,colors):
    pct=val/sum(vals)*100
    ax.barh(["Apps"],pct,left=left,color=color,height=0.5,edgecolor="white",linewidth=1.5,label=cat)
    if pct>5: ax.text(left+pct/2,0,f"{pct:.0f}%",ha="center",va="center",fontsize=7,fontweight="bold",color="white")
    left+=pct
ax.set_xlim(0,100); ax.legend(fontsize=6,ncol=6,loc="upper center",bbox_to_anchor=(0.5,-0.3))
ax.set_title("100% Stacked Bar: The Better Alternative to Pie",fontweight="bold")
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
plt.tight_layout(); plt.savefig("output/stacked_100_py.png",dpi=300); plt.close()

print("All W03-M05 Python plots saved")
